drivers/gnss/gnss_nmea0183.c

```diff
@@ -0,0 +1,678 @@
+	if (argc < 2) {
+		return false;
+	}
+
+	/* First argument should start with '$' which is not covered by checksum */
+	if ((argc < 1) || (argv[0][0] != '$')) {
+		return false;
+	}
+
+	len = strlen(argv[0]);
+	for (uint16_t u = 1; u < len; u++) {
+		checksum ^= argv[0][u];
+	}
+	checksum ^= ',';
+
+	/* Cover all except last argument which contains the checksum*/
+	for (uint16_t i = 1; i < (argc - 1); i++) {
+		len = strlen(argv[i]);
+		for (uint16_t u = 0; u < len; u++) {
+			checksum ^= argv[i][u];
+		}
+		checksum ^= ',';
+	}
+
+	if ((gnss_parse_atoi(argv[argc - 1], 16, &tmp) < 0) ||
+	    (tmp > UINT8_MAX) ||
+	    (tmp < 0)) {
+		return false;
+	}
+
+	return checksum == (uint8_t)tmp;
+}
+
+int gnss_nmea0183_knots_to_mms(const char *str, int64_t *mms)
+{
+	int ret;
+
+	__ASSERT(str != NULL, "str argument must be provided");
+	__ASSERT(mms != NULL, "mms argument must be provided");
+
+	ret = gnss_parse_dec_to_nano(str, mms);
+	if (ret < 0) {
+		return ret;
+	}
+
+	*mms = (*mms) / GNSS_NMEA0183_NANO_KNOTS_IN_MMS;
+	return 0;
+}
+
+int gnss_nmea0183_parse_hhmmss(const char *hhmmss, struct gnss_time *utc)
+{
+	int64_t i64;
+	int32_t i32;
+	char part[3] = {0};
+
+	__ASSERT(hhmmss != NULL, "hhmmss argument must be provided");
+	__ASSERT(utc != NULL, "utc argument must be provided");
+
+	if (strlen(hhmmss) < 6) {
+		return -EINVAL;
+	}
+
+	memcpy(part, hhmmss, 2);
+	if ((gnss_parse_atoi(part, 10, &i32) < 0) ||
+	    (i32 < 0) ||
+	    (i32 > 23)) {
+		return -EINVAL;
+	}
+
+	utc->hour = (uint8_t)i32;
+
+	memcpy(part, &hhmmss[2], 2);
+	if ((gnss_parse_atoi(part, 10, &i32) < 0) ||
+	    (i32 < 0) ||
+	    (i32 > 59)) {
+		return -EINVAL;
+	}
+
+	utc->minute = (uint8_t)i32;
+
+	if ((gnss_parse_dec_to_milli(&hhmmss[4], &i64) < 0) ||
+	    (i64 < 0) ||
+	    (i64 > 59999)) {
+		return -EINVAL;
+	}
+
+	utc->millisecond = (uint16_t)i64;
+	return 0;
+}
+
+int gnss_nmea0183_parse_ddmmyy(const char *ddmmyy, struct gnss_time *utc)
+{
+	int32_t i32;
+	char part[3] = {0};
+
+	__ASSERT(ddmmyy != NULL, "ddmmyy argument must be provided");
+	__ASSERT(utc != NULL, "utc argument must be provided");
+
+	if (strlen(ddmmyy) != 6) {
+		return -EINVAL;
+	}
+
+	memcpy(part, ddmmyy, 2);
+	if ((gnss_parse_atoi(part, 10, &i32) < 0) ||
+	    (i32 < 1) ||
+	    (i32 > 31)) {
+		return -EINVAL;
+	}
+
+	utc->month_day = (uint8_t)i32;
+
+	memcpy(part, &ddmmyy[2], 2);
+	if ((gnss_parse_atoi(part, 10, &i32) < 0) ||
+	    (i32 < 1) ||
+	    (i32 > 12)) {
+		return -EINVAL;
+	}
+
+	utc->month = (uint8_t)i32;
+
+	memcpy(part, &ddmmyy[4], 2);
+	if ((gnss_parse_atoi(part, 10, &i32) < 0) ||
+	    (i32 < 0) ||
+	    (i32 > 99)) {
+		return -EINVAL;
+	}
+
+	utc->century_year = (uint8_t)i32;
+	return 0;
+}
+
+int gnss_nmea0183_parse_rmc(const char **argv, uint16_t argc, struct gnss_data *data)
+{
+	int64_t tmp;
+
+	__ASSERT(argv != NULL, "argv argument must be provided");
+	__ASSERT(data != NULL, "data argument must be provided");
+
+	if (argc != 15)  {
+		return -EINVAL;
+	}
```

Comment:
hey had a go at writing a generic uart "dumb" driver to play with the API, this one fails cause I only get 14 fields:

`[00:03:45.286,346] <dbg> modem_chat: modem_chat_log_received_command: $GNRMC, 145732.00 A 4214.65438 N 00158.45841 E 0.072  250923   A 6D`

this some random ebay GPS module that claims to be a `UBX-M80xx`, looks like it's missing the last `V`. Looking for format around the net it seems like this is a bit all over the place, should this check be relaxed?

Hacked a 14 here and it worked pefectly.

Reply:
I have relaxed these checks to only cover the number of arguments we actually check :)

Reply:
makes sense, thanks!

Reply:
Also, awesome that it seems to just work!